heartbeat/monitors/factory.go

```diff
@@ -78,8 +91,45 @@ func (f *RunnerFactory) Create(p beat.Pipeline, c *common.Config) (cfgfile.Runne
 	}
 
 	p = pipetool.WithClientConfigEdit(p, configEditor)
-	monitor, err := newMonitor(c, plugin.GlobalPluginsReg, p, f.sched)
-	return monitor, err
+
+	f.mtx.Lock()
+	defer f.mtx.Unlock()
+
+	// This is a callback executed on stop of a monitor, it ensures we delete the entry in
+	// byId.
+	// It's a little tricky, because it handles the problem of this function being
+	// occasionally invoked twice in one stack.
+	// f.mtx would be locked given that golang does not support reentrant locks.
+	// The important thing is clearing the map, not ensuring it stops exactly on time
+	// so we can defer its removal from the map with a goroutine, thus breaking out of the current stack
+	// and ensuring the cleanup happen soon enough.
+	safeStop := func(m *Monitor) {
+		go func() {
+			// We can safely relock now, since we're in a new goroutine.
+			f.mtx.Lock()
+			defer f.mtx.Unlock()
+
+			// If this element hasn't already been removed or replaced with a new
+			// instance delete it from the map. Check monitor identity via pointer equality.
+			if curM, ok := f.byId[m.stdFields.ID]; ok && curM == m {
+				delete(f.byId, m.stdFields.ID)
+			}
+		}()
+	}
+	monitor, err := newMonitor(c, f.pluginsReg, p, f.sched, safeStop)
+	if err != nil {
+		return nil, err
+	}
+
+	if mon, ok := f.byId[monitor.stdFields.ID]; ok {
+		f.logger.Warnf("monitor ID %s is configured for multiple monitors! IDs should be unique values, last seen config will win", monitor.stdFields.ID)
+		// Stop the old monitor, since we'll swap our new one in place
+		mon.Stop()
```

Comment:
Can this cause scheduling problems? For example if a monitor is scheduled every 30s, but for some reason an identical new one is created every 10s, replacing the old one, is it possible that it ends up being executed every 10s (or never?).

Reply:
Great question. Fortunately that won't happen! The monitor object is bound to the scheduler [here](https://github.com/elastic/beats/blob/master/heartbeat/monitors/task.go#L87), so the way this works should be fine.

Reply:
Oh, I see your point, I guess my question is why would an identical new one be created every 10s? We do run monitors immediately on create. I think, in that situation, we need to explore why a new one, with a different config, is created every 10s. This patch isn't a substitute for config de-duplication I would say.

Reply:
> why would an identical new one be created every 10s?

It would be quite a corner case, in autodiscover for example if something that triggers autodiscover events is being periodically updated. Though in this case the issue would be there also without this change, as on update the config is already stopped an started.
So we are probably fine as this is now.